src/Uno.UI/UI/Xaml/DependencyObjectStore.cs

```diff
@@ -848,40 +898,69 @@ internal IDisposable RegisterParentChangedCallback(object key, ParentChangedCall
+				ParentChangedCallback weakCallback =
+					(s, _, e) => (weakCallbackRef.Target as ParentChangedCallback)?.Invoke(s, key, e);
 
-				_parentChangedCallbacks = _parentChangedCallbacks.Add(weakDelegate);
+				_parentChangedCallbacks = _parentChangedCallbacks.Add(weakCallback);
 
 				// This weak reference ensure that the closure will not link
 				// the caller and the callee, in the same way "newValueActionWeak"
 				// does not link the callee to the caller.
 				var instanceRef = ThisWeakReference;
 
-				void Cleanup()
-				{
-					var that = instanceRef.Target as DependencyObjectStore;
+				return new RegisterParentChangedCallbackConditionalDisposable(
+					instanceRef.CloneWeakReference(),
+					instanceRef,
+					weakCallbackRef,
+					weakCallback,
+					callback
+				);
+			}
+		}
 
-					if (that != null)
-					{
-						// Delegates integrate a null check when removing new delegates.
-						that._parentChangedCallbacks = that._parentChangedCallbacks.Remove(weakDelegate);
-					}
+		/// <summary>
+		/// Specialized DispatcherConditionalDisposable for <see cref="RegisterParentChangedCallback(object, ParentChangedCallback)"/>
+		/// </summary>
+		/// <remarks>
+		/// This class is used to avoid the creation of a set of <see cref="Action"/> instances, as well as delegate invocations.
+		/// </remarks>
+		private class RegisterParentChangedCallbackConditionalDisposable : DispatcherConditionalDisposable
+		{
+			private ManagedWeakReference _doStoreRef;
+			private ManagedWeakReference _weakCallbackRef;
+			private ParentChangedCallback _weakCallback;
+			private ParentChangedCallback _callback;
 
-					WeakReferencePool.ReturnWeakReference(that, wr);
+			public RegisterParentChangedCallbackConditionalDisposable(
+				WeakReference conditionSource,
+				ManagedWeakReference doStoreRef,
+				ManagedWeakReference weakCallbackRef,
+				ParentChangedCallback weakCallback,
+				ParentChangedCallback callback) : base(callback.Target, conditionSource)
+			{
+				_doStoreRef = doStoreRef;
+				_weakCallbackRef = weakCallbackRef;
+				_weakCallback = weakCallback;
+				_callback = callback;
+			}
 
-					// Force a closure on the callback, to make its lifetime as long
-					// as the subscription being held by the callee.
-					callback = null!;
+			protected override void DispatchedTargetFinalized()
+			{
+				var that = _doStoreRef.Target as DependencyObjectStore;
+
+				if (that != null)
+				{
+					// Delegates integrate a null check when removing new delegates.
+					that._parentChangedCallbacks = that._parentChangedCallbacks.Remove(_weakCallback);
 				}
 
-				return new DispatcherConditionalDisposable(
-					callback.Target,
-					instanceRef.CloneWeakReference(),
-					Cleanup
-				);
+				WeakReferencePool.ReturnWeakReference(that, _weakCallbackRef);
+
+				// Force a closure on the callback, to make its lifetime as long
+				// as the subscription being held by the callee.
+				_callback = null!;
```

Comment:
Is there a reason for `null!` instead of declaring the field nullable?

Reply:
There isn't anymore. It used to be a non-nullable parameter lifted in a lambda, now that the variable can be null as it's a field in a class.

Will fix in another PR.